lib/collection/src/shards/resharding/driver.rs

```diff
@@ -363,3 +487,79 @@ fn stage_propagate_deletes() -> CollectionResult<()> {
 fn stage_finalize() -> CollectionResult<()> {
     todo!()
 }
+
+/// Await for a resharding shard trnasfer to succeed.
+///
+/// Yields on a successfull transfer.
+///
+/// Returns an error if:
+/// - the transfer failed or got aborted
+/// - the transfer timed out
+/// - no matching transfer is ongoing; it never started or went missing without a notification
+///
+/// Yields on a succesful transfer. Returns an error if an error occurred or if the global timeout
+/// is reached.
+async fn await_transfer_success(
+    reshard_key: &ReshardKey,
+    transfer: &ShardTransfer,
+    shard_holder: &Arc<LockedShardHolder>,
+    collection_id: &CollectionId,
+    consensus: &dyn ShardTransferConsensus,
+    await_transfer_end: impl Future<Output = CollectionResult<Result<(), ()>>>,
+) -> CollectionResult<()> {
```

Comment:
Though this and it's usage may seem a bit complex, there's three important things it does:

1. we subscribe to shard transfer changes before we actually start transfers, to ensure we don't miss notifications
2. it enforces a timeout, we don't wait longer than 24 hours for a transfer to complete
3. it does a periodic sanity check to ensure the transfer is still active in consensus, if it isn't we assume it has failed (and panic in debug mode because this cannot happen)